### ISSUE TITLE
Optional check suggestion: prefix function declarations with `function`

### ISSUE THREAD

**Issue by felipecrs** [koalaman/shellcheck#2645]
This is to ensure a consistent code style. Example:

```bash
myfunction() {
  true
}
```

Should be:

```bash
function myfunction() {
  true
}
```

**Comment by wolfspyre** [comment]
I came here to actually suggest the opposite of this. 

according to [This SO post](https://stackoverflow.com/questions/22238033/should-i-define-a-shell-function-as-function-x-or-just-x);  the use of:

 `function myfunction() { ... ; }` 

is completely wrong, and shouldn't be used; and that the expected syntax is:

`myfunction(){ ...; }`

 
I'm not confident in this, but would enjoy seeing the resulting conversation around correctness here ;)

**Comment by wileyhy** [comment]
This seems like a "try it and see" kind of question to me. :-)

~ $ bash --version
GNU bash, version 5.3.0(1)-release (aarch64-unknown-linux-android)
~ $ function bla2 () { printf '<%s>\n' "$@";}
~ $ bla2 4 5 76
<4>
<5>
<76>
~ $ unset -f bla2
~ $

Perhaps this 'function word ()...' syntax causes an error in some other
shell?

**Comment by wolfspyre** [comment]
erm... no? 

Just cause somethin works doesn't mean it's necessarily a good idea :) 

according to the docs 'function' AND '()' aughtn't be used in conjunction. bash and zsh enable it, but it's not proper, strictly speakin....
will it work? sure. but if that's all one cared about, why would one use shellcheck in the first place? :) 

Since we're talking about a tool which is intended to suggest patterns which reflect current best practices I'd hope that effaroundandfindout isn't the de-facto mechanism of evaluating the merit of a pattern... 

:)

but I do admit this is kinda nit-picky...

**Comment by felipecrs** [comment]
I agree with you, @wolfspyre, `f() { echo hello world; }` should be the preferred one. I updated the issue description to make this clearer.

Also referenced in https://mywiki.wooledge.org/Bashism.

**Comment by wileyhy** [comment]
Well, there are some varying levels of "proper" under consideration.

Would 'function word ()...' conform to [pick your favorite corporation's]
Style Guide? Maybe / maybe not. That's up to them.

Is said syntax a shell error, meaning, could such syntax cause, for instance
  - a backup script to silently fail,
  - data to be read or processed incorrectly,
  - unexpected side effect(s) to occur, or
  - for one to "incur the wrath from high atop the thing?"

Case in point:

  <sh>
  ~ $ type -a '!'
  ! is a shell keyword
  ~ $ function '!' () { printf '%d\n' "${LINENO}";}
  bash: `'!'': not a valid identifier
  ~ $
  <\sh>

In my opinion, if people claim some syntax is "wrong," it may be worthwhile
to ask for whom and when.

**Comment by felipecrs** [comment]
That's why this ask is for an optional check, like these:

- https://www.shellcheck.net/wiki/SC2250
- https://www.shellcheck.net/wiki/SC2248
- https://www.shellcheck.net/wiki/SC2292

**Comment by brother** [comment]
I would like a optional check for this indeed. (on team "don't use the function keyword"  =))
